### ISSUE TITLE
Support specifying that two nodes should be at the same level/rank

### ISSUE THREAD

**Issue by CAFxX** [mermaid-js/mermaid#3723]
This is basically a clone of https://github.com/mermaid-js/mermaid/issues/418 that was automatically closed as stale without resolution.

**Is your feature request related to a problem? Please describe.**

I need a way to specify that two or more nodes should be placed at the same level/rank 

For example, in the following graph I would like `b` and `c` to be on the same level

```
graph TD
a --> b & c
b --> d & c
c --> e & b
```

this currently yields something like the following graphs (notice how `b` and `c` are not at different heights

![image](https://user-images.githubusercontent.com/408018/197999331-aef54dee-50d1-461b-89fd-d12b5e97af4e.png) 

**Describe the solution you'd like**

 I would like `b` and `c` to behave like in this diagram 

```
graph TD
a --> b & c
b --> d
c --> e
```

![image](https://user-images.githubusercontent.com/408018/197998812-05f07342-080d-4d7a-a00d-21dc43ecabde.png)

**Describe alternatives you've considered**

graphviz allows to do this by specifying [`rank=same`](https://graphviz.org/docs/attrs/rank/)

**Additional context**

**Comment by SpacyRicochet** [comment]
To take this one step further, I would love it to see if this also applies to subgraphs themselves.

```text
flowchart TD

A
subgraph S1; B; C; end
subgraph S2; D; E; end
B --> C
A -- YES --> B
A -- NO  --> D
C -- YES --> D
C -- NO  --> F
D --> E
E --> F
```

![mermaid-diagram-2022-10-26-124710](https://user-images.githubusercontent.com/904596/198007158-f803dd30-ff22-4a29-aa8d-24ac84230058.svg)

It would be great if the subgraphs here could be on the same rank, allowing A to be a very obvious start, and F a very obvious end.

```text
┌───────────────────────────────────┐
│               ┌───┐               │
│    ┌──────────┤ A │───┐           │
│  YES          └───┘  NO           │
│    │                  │           │
│    │                  │           │
│    │                  │           │
│ ┌──┼─────────┐     ┌──┼─────────┐ │
│ │  │  S1     │     │  │  S2     │ │
│ │  ▼         │     │  ▼         │ │
│ │┌───┐  ┌───┐│     │┌───┐  ┌───┐│ │
│ ││ B │─▶│ C │├─YES─▶│ D │─▶│ E ││ │
│ │└───┘  └───┘│     │└───┘  └───┘│ │
│ │         │  │     │         │  │ │
│ │         │  │     │         │  │ │
│ └─────────┼──┘     └─────────┼──┘ │
│           │                  │    │
│          NO                  │    │
│           │                  │    │
│           │   ┌───┐          │    │
│           └──▶│ F │◀─────────┘    │
│               └───┘               │
└───────────────────────────────────┘
```

**Comment by reedsutliff** [comment]
would love to see this feature as well.

**Comment by olivierduquenoy** [comment]
Yes, it will be a great improvement !!

**Comment by brandond** [comment]
Here for a +1 on the ability to place links between nodes of the same rank.

Just using the sample cross-connected diagram, if I want to connect the two nodes as peers, it goes from something fairly intuitive, into complete garbage, because the layout code demotes the B node down to rank 2, and C/D down to rank 3, because of the link from A to B.

```
graph TD
  A --> C
  A --> D
  B --> C
  B --> D
```

![image](https://user-images.githubusercontent.com/370103/226489661-d21c6040-9670-43f5-8f15-49a5cb07fe86.png)

```
graph TD
  A --> C
  A --> D
  B --> C
  B --> D

  A o--o B
```

![image](https://user-images.githubusercontent.com/370103/226489812-b09df99c-89df-4f7e-b71d-bac637d3caa8.png)

**Comment by AartBluestoke** [comment]
trying to show the relationship between 2 groups of objects. the group subgraphs can't be put at the same height, one is "higher" than the other.

**Comment by ignamiranda** [comment]
I would also really like this feature

**Comment by ignamiranda** [comment]
Adding on to this, I understand why `A --> B` would have a parent-child relationship but I don't understand why `A <--> B` or `A --- B` would. It'd be a departure from the way it's been a long time, but having non hierarchical, horizontal links could be a cleaner, more intuitive solution than having a flag to declare two nodes as same rank.

**Comment by nickovs** [comment]
@ignamiranda I think that we need the ability to tag nodes as having the same rank even when they don't have a link, for a couple of reasons.
 * Firstly, aside from breaking many existing diagrams, making symmetrical links force the same rank will make it much harder to control the layout of undirected graphs. There are plenty of cases where the data represents an undirected graph but you want to control the presentation anyway. 
 * Secondly, there are scenarios where we may want to place two nodes on the same rank but not draw a link between them and not necessarily have them closer together. While in your model you could use `A ~~~ B` to equalise the rank, doing that still places a "spring" between the nodes in the layout process. Being able to explicitly place two nodes on the same rank without forcing those nodes closer together would be helpful, and the intent would probably be clearer too.

**Comment by brandond** [comment]
Is there any consensus on the approach here, such that someone could actually pick this up and work on it? I keep stumbling over this in our flow diagrams.

**Comment by brandond** [comment]
cc @sidharthv96 @knsv

**Comment by eschalkargans** [comment]
Just updating the thread as not being able to rank nodes is a major limitation of mermaid diagrams.

Example: In a Top->Bottom diagram, I might need  some nodes are aligned on row 1, row 2, row 3, etc, forcing the layout

The `~~~` syntax does not work.

Intuitively, I expected that subgraphs would allow such features, but the documentation itself says that when edges are coming out of a subgraph, alignments won't be applied

It would be such a nice feature to be able to rank node!

**Comment by cmdrrickhunter2** [comment]
The lack of this feature has forced me to use other tools for rendering category theory graphs.  One of the most fundamental structures in category theory is the commuting square.  This square is represented _exactly_ the same in every category theory reference I've ever come across:

    A   -- f -->  B
    |             |
    g             g'
    |             |
    v             v
    A'  -- f' --> B'

Hence the term commuting _square_.  Mermaid renders this as

```mermaid
%%{ init: {'flowchart': {'curve': 'linear' } } }%%
    graph LR;
      A -- f --> B
      A' -- f' --> B'
     A -- g --> A'
     B -- g' --> B'
```

which has the same connectivity, but lacks the instant recognition as a square.  If I could render this the same way I can render
```mermaid
    graph LR;
      A -- f --> B
      A' -- f' --> B'
``` 
then the result would be usable.  But category theory is far too complicated to display the graphs in a non-standard way.  The graphs need to clarify the math that goes behind them.  Recognizable graph patterns are essential.

**Comment by brandond** [comment]
Yeah, it's unfortunate that mermaid lacks support for something that seems so basic.

**Comment by drk-mtr** [comment]
This would be a great feature. Unfortunately without it simple charts are non-viable for basic cases.

If I have a top-down chart `A --> B --> C --> D` and the `C` and `D` nodes fall off the bottom of the page, I need some way to say "display C at the same rank as A", else mermaid cannot be used as the solution.

**Comment by Spiralo-Idioide** [comment]
This is still something that seems would be very useful for a lot of use cases. For the moment making any kind of loops is awful  (something that is important ex: #2493)

**Comment by hginzel** [comment]
Consider this simple example for same rank, please.
```
graph LR
A --> D
B --> C[Intermediate] --> D
```
```mermaid
graph LR;
A --> D
B --> C[Intermediate] --> D
```

**Comment by AartBluestoke** [comment]
> Consider the simple example for [_same rank_](https://graphviz.org/docs/attrs/rank/) for nodes `A` and `B`, please.
> 
> ```
> graph LR;
> A --> D
> B --> C[Intermediate] --> D
> ```
> 
> 
> It seems this particular graph can be solved using _longer edge_.
> 
> ```
> graph LR;
> %% one more dash here
> A ---> D
> B --> C[Intermediate] --> D
> ```

@hginzel 

The core of the issue is you can't have both vertical and horizontal lines at the same time. attempts to put those in forces a major re-alignment of the layout.

**Comment by sageacy** [comment]
+1 to an alignment feature.

I would argue that a "longer edge" is not alignment; it may allow you to tweak things to force alignment, but ideally, the alignment would be declarative. The 'rank' solution in Graphviz is a reasonable approach!

**Comment by clocksmith** [comment]
I agree with @sageacy 's suggestion for a syntax-level solution to handle the "same rank" issue, similar to the approach used by GraphViz.

The underlying mathematical algorithms for this type of alignment have been well-established for decades. Therefore, the main challenge is designing the public API. A declarative alignment syntax seems like the most promising approach, and implementing it would also be a great opportunity to improve the default layout behavior.

**Comment by njelich** [comment]
Ditte this! It is a really useful feature.

**Comment by jimtendo** [comment]
+1 for this feature.

Does anyone know if MermaidJS Flowcharts use the concept of `rank` under-the-hood?

I'm trying to work out what the dev-effort here would be like and whether it's mostly just a matter of defining a syntax that allows rank to be specified for the given nodes.

Also, does anyone have suggestions for a syntax for doing this?

**Comment by AartBluestoke** [comment]
> Also, does anyone have suggestions for a syntax for doing this?

A ~~~ B tries to set A and B to be at the same rank, it is an otherwise invisible link? - this is the most backwards compatable way
symmetrical links that don't force A or B to be lower, even though at least a --- b currently forces b to be a rank lower than A
A <--> B 
A ---- B

**Comment by nickovs** [comment]
@AartBluestoke While there's an existing syntax for an invisible link, as [noted above](https://github.com/mermaid-js/mermaid/issues/3723#issuecomment-1657234623) this has the side effect of also adding a layout 'spring' between the nodes that encourages the nodes to be moved closer together, so it does more than just align the ranks.

**Comment by AndreKR** [comment]
I think this could be solved by having a way to specify that an edge should not affect rank at all? This would allow the user to tweak the node layout using invisible edges (if that is even necessary) and then draw additional edges without disturbing the node layout.

**Comment by awwaiid** [comment]
OK well, I gave it a go to make a step toward this and it initially seems like it is working. I didn't invent new syntax (yet), but instead made it so you can add a `constraint: false` attribute to a named edge. Like:

```
flowchart TB
  A --> B
  B e1@--> C
  e1@{ constraint: false }
```

If this is the right direction then I can follow any missing PR contribution guidelines and submit it. The change was small enough that I assume it falls apart in complex cases, but I did a few simple ones and uploaded to a gist.

* [demo](https://gistpreview.github.io/?5b7303a1faf638cd5ff0afc6f7a7d145)
* [for with impl](https://github.com/awwaiid/mermaid)

**Comment by knsv** [comment]
Funny, I was just mentioned this concept in #6817 

What you are describing is very much in line with what Mermaid needs. It will be tightly coupled with the layout algorithm being used, though. One directive, layer, closeTo, rightOf, or similar might not be applicable in all algorithms. Layer for instance will work well in dagre but not in an orthogonal layout.

**Comment by awwaiid** [comment]
I have this working for dagr, I'll check but I'm guessing other renderers would ignore the flag and render as is by default, which seems good.

**Comment by knsv** [comment]
Thats fantastic @awwaiid!

Maybe create a draft PR where we can sort out the details and syntax etc?

**Comment by awwaiid** [comment]
I followed the contributor guide and PR template and submitted https://github.com/mermaid-js/mermaid/pull/7463!
